README
MVC-Neaf

MVC::Neaf [ni:f] stands for Not Even A (Web) Framework.

See README.md or `perldoc lib/MVC/Neaf.pm` for feature discussion.

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc MVC::Neaf

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=MVC-Neaf

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/MVC-Neaf

    CPAN Ratings
        http://cpanratings.perl.org/d/MVC-Neaf

    Search CPAN
        http://search.cpan.org/dist/MVC-Neaf/


LICENSE AND COPYRIGHT

Copyright (C) 2016-2018 Konstantin S. Uvarin <khedin@cpan.org>

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
# Revision history for MVC::Neaf

0.29    Fri Jul 28 2023
    - Calculate template & static content paths relative to the calling file,
    -   not current directory.
    - Allow to put get/post/... handler after, not before other options.

0.28    Mon Oct 21 2019
    - TST Increase test coverage + some refactoring

0.2701  Thu May 30 2019
    - BUG Fix regex type check on perls <= 5.010

0.27    Thu May 30 2019
    - API call load_resources(\*DATA) upon run() automatically
    - API Add body_json(), body_raw(), body_text(), and content_type() calls to request
    - API custom error handlers (neaf 404) can be configure per-path
    - API BREAK Replace a JSON default error page with a HTML one
    - API Add magic(bool) switch
    - API Add regexp check for headers

0.2603  Sat Nov 24 2018
    - BUG Fix Apache2 test warning

0.2602  Sat Nov 24 2018
    - BUG Fix (?) failing test

0.2601  Fri Jun 22 2018
    - BUG Skip empty parameters in strict mode

0.26    Sun Jun 10 2018
    - API BREAK rework load_resources, @@ foo.tpl view=TT now supported
    - TST Improve coverage, add tests, refactor things

0.2501  Sun Jun 3 2018
    - API BREAK Restrict characters for path, in particular ":" (a warning only)
    - BUG Understand double slashes in URLs
    - BUG Fix routing bug - was treating /pathogen as /path

0.25    Sat Jun 2 2018
    - API Add 'strict' flag that makes controller emit error 422 if params/cookies pass validation
    - API Add overridable log_message, make_id to Request class
    - API BREAK Remove deprecated error_template, set_default, server_stat
    - API BREAK Remove Request->upload (use utf8/raw), Request->set_full_path (use set_path)
    - REF Use base64url instead of base64 everywhere
    - NET BREAK Deprecate Apache2 backend in favor of Plack::Handler::Apache2
    - MOD BREAK Remove ServerStat

0.24    Sat May 26 2018
    - API Add 'helper' subs - user-defined, path-restricted $req methods
    - API BREAK Rename -serial => -payload in app response
    - API set_path_defaults can be restricetd on path + method
    - REF Rename MVC::Neaf::Route::{Recursive=>Main}
    - REF Util::Container - path&method based container util
    - REF Add Route::PreRoute stub class for hooks & helpers

0.23    Wed May 23 2018
    - API Add Route class, add $req->route() call for showing current route
    - API add prefix(), postfix(), and splat() for URI path components
    - API Make :sugar keyword optional, export get+post & neaf by default
    - API BREAK deprecate MVC::Neaf->something() (expect new() of course)
    - API BREAK Kill deprecated neaf->pre_route for good
    - API Add -headers support to exception
    - REF A giant refactoring of the core
    - BUG A lot of bugs introduced & then fixed

0.2203  Thu Jan 25 2018
    - AUX Add github links to Makefile.PL

0.2202  Fri Jan 19 2018
    - TST Fix tests under Plack < 1.043 https://rt.cpan.org/Ticket/Display.html?id=124124

0.2201  Thu Jan 18 2018
    - DOC Fix escaping in docs

0.22    Thu Jan 18 2018
    - DOC Rearrange core doc sections
    - DOC Deprecate X::Stat module, say replacement pending
    - API BREAK Deprecate server_stat method
    - API BREAK Reprecate req->header_in_keys
    - BUG Work around HTTP::XSHeaders not providing flatten* methods
    - TST Add a performance statistics example
    - TST Fix(?) tests on older perls

0.21    Thu Dec 7 2017
    - API Add any, put, patch, and del for DELETE to sugar
    - API route(): Add 'public', 'tentative', 'override' flags
    - MOD static '/path' => [ "content" ] to serve in-memory files
    - TPL Add View::TT->preload for in-memory templates
    - API Experimental load_resources for Mojo-like @@ files in __DATA__
    - TPL Minor fixes in View::JS, add want_pretty and want_sorted options
    - DEV Add callback(\%, path, method) to get_routes
    - REF unify die 404 with MVC::Neaf::Exception
    - REF Use HTTP::Headers::Fast everywhere
    - BUG Tighten dependencies, add more workarounds for failing deps
    - BUG Fix error reporting in diamond upload test
    - TST Improve test suite somewhat

0.2002  Fri Dec 1 2017
    - BUG Fixed escapes in POD

0.2001  Fri Dec 1 2017
    - Version bump NOCODE

0.20    Fri Dec 1 2017
    - API Add 'neaf form => name => {...}' for kosher input validation
    - MOD Add X::Form::Wildcard module to handle unforeknown request params
    - API Add public log_error() method to Request, use it for all errors
    - API BREAK Split upload() into upload_utf8() and upload_raw()
    - API BREAK Remove req->set,get_default for good
    - REF Replace JSON dependency with JSON::MaybeXS
    - REF Replace concatenation with md5 in req->id generator (todo SHA3)
    - DOC Improve Neaf documentation NOCODE
    - BUG Multiple bugs fixed, new ones added

0.1901  Tue Nov 28 2017
    - BUG Fix local $SIG{__WARN__} in BEGIN section

0.19    Fri Nov 24 2017
    - CLI Add --listen option to CLI interface
    - API Add exp. path_info_split method to fetch regex capture groups
    - API BREAK default error = json + logged err_id
    - API BREAK Deprecate foo => bar => baz => sub {}-style paths in route()
    - API BREAK Remove CGI driver for good, remove CGI.pm dependency
    - BUG Add hack to prevent tests from failing on older perls

0.18    Mon Nov 20 2017
    - TST Rework examples to use the new declarative get+post syntax
    - API BREAK Use -view=>JS by default; switch to TT & issue warning if -template seen
    - API Add experimental predefined param_regex to route
    - TPL Add non-hash response support to JS view via -serial return key
    - DEV Add req->endpoint_origin method for file:line where current handler was defined
    - API BREAK Deprecate set_full_path() for good
    - DEV run_test(): add support for body, hash cookies, https
    - CLI More details in --list output
    - REF A BIG refactoring of core, handle_request() broken down
    - BUG Lots of bug fixes

0.17    Fri Apr 28 2017

Makefile.PL
use 5.006;
use strict;
use warnings;
use ExtUtils::MakeMaker;

my $maker = ExtUtils::MakeMaker->VERSION;

my %test_req = (
        'Test::More'          => 0.87, # done_testing
        'Test::Warn'          => 0,
        'File::Temp'          => 0,
);

WriteMakefile(
    NAME                 => 'MVC::Neaf',
    AUTHOR               => q{Konstantin S. Uvarin <khedin@gmail.com>},
    VERSION_FROM         => 'lib/MVC/Neaf.pm',
    ABSTRACT_FROM        => 'lib/MVC/Neaf.pm',
    ($maker >= 6.3002
      ? ('LICENSE'=> 'perl')
      : ()),
    MIN_PERL_VERSION     => 5.008,
    PL_FILES             => {},
    PREREQ_PM => {
        $] >= 5.01 ? () : (
            'Digest::SHA'     => 0,    # TODO use SHA3 instead
            'parent'          => 0,    # not present in 5.8
        ),
        'ExtUtils::MakeMaker' => 6.64, # for TEST_REQUIRES
        'HTTP::Headers::Fast' => 0.21, # psgi_flatten_without_sort
        'JSON::MaybeXS'       => 0,    # also by Plack via HTTP::Entity::Parser
        'Plack::Request'      => 0,
        'Template'            => 0,
        'URI::Escape'         => 0,
        $maker >= 7.12 ? () : (        # buggy test_requires, so emulate
            %test_req,
        ),
    },
    TEST_REQUIRES => {
        %test_req,
    },
    META_MERGE => {
        "meta-spec" => { version => 2 },
        prereqs => {
            develop => {
                requires => {
                    'DBD::SQLite'    => 0, # sql-based session
                    'DBI'            => 0,
                    'GD::Simple'     => 0, # for png example
                    'LWP::UserAgent' => 0, # it/*.t depends on it
                    (($] < 5.010) ? () : ('Validator::LIVR' => 0)),
                }
            },
            runtime => {
                recommends => {
                    'JSON::XS' => 0,
                    'Validator::LIVR' => 0,
                },
            },
        },
    },
    dist                 => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean                => { FILES => 'MVC-Neaf-*' },
);
